### ISSUE TITLE
Using getter to calculate computed values

### ISSUE THREAD

**Issue by filipesmedeiros** [react-spring/zustand#132]
Hey!

Is it ok to use the getter to return a computed value in a function? Since using it in a value doesn't work?

Example:

```typescript
const [useAuth] = create((set, get) => ({
    user: {username: undefined, authLevel: 0},
    // isSignedIn: !!get().user.username // doesn't work because it's executed before the store is created, so get() returns undefined right?
    isSignedIn: () => get().user.username // should work, just call isSignedIn() instead of using as a value, in the components?
})}
```

Also, I feel like computed values are a cool features for state management libraries to have out of the box, are there planes to add it? Do you guys need help with it somehow?

**Comment by shahruz** [comment]
I usually would just do something like `const isSignedIn = useAuth(state => !!state.user.username);` in the consuming component for this.

**Comment by filipesmedeiros** [comment]
Right! That seems prettier. But that only works for component use right? If I want to separate some logic while calculating various layers of computed values in a store (in more complex cases) that doesn't work anymore. For example:

```typescript
const [useStore] = create((set, get) => {
    value: 30,
    value1: get().value + 60,
    value2: get().value1 / 4,
    value3: get().value2 * 1000
})
```

Logically, it might be better if `value3` depends on `value2` directly than if you had to repeat the computations you did for the other ones, right?

**Comment by jordanmkoncz** [comment]
I was also curious about the recommended approach for selectors which derive/compute values from state (and possibly from other selectors themselves). Reading about Recoil and how it handles this using [selectors](https://recoiljs.org/docs/basic-tutorial/selectors) (also see https://www.youtube.com/watch?v=_ISAA_Jt9kI), I couldn't see anything analogous in Zustand which has the same optimisations.

**Comment by drcmda** [comment]
sure you can use getters like that.

as for recoil and atoms, it's a slightly different paradigm. we have something coming up that's concentrating on that approach.

**Comment by dai-shi** [comment]
@drcmda I don't think getters work because of `Object.assign()`. Once you `setState({ user })`, `signedIn` becomes a boolean value. I didn't try it, so correct me if I'm wrong.

**Comment by jordanmkoncz** [comment]
Thanks @dai-shi for your response.

>b) Chaining computed values (aka atoms). For b), we are planning a new project coming soon, and it's outside the scope of zustand.

Just to confirm, you're saying that chaining computed values in a similar fashion to Recoil and how it handles selectors (where computed values are automatically re-computed when one of their 'dependencies' changes, and *only* re-computed when this happens) is not something you plan to add to Zustand itself, but instead you are planning on creating a new project that will address this and can be used with Zustand?

**Comment by dai-shi** [comment]
Hi @jordanmkoncz 

> chaining computed values is not something you plan to add to Zustand itself

Correct. I think that will be overkill for the Zustand core.
It would be possible for middleware, though.

> instead you are planning on creating a new project that will address this and can be used with Zustand?

Yes for the first part and no for the second. The new project is something different from Zustand, but with the simplicity philosophy from Zustand.

**Comment by timkindberg** [comment]
Another option is pure functions:

```js
const useAuth = create((set, get) => ({
    user: {username: undefined, authLevel: 0}
})}

const select = {
    signedIn: state => state.user.username
}

// in components
function Foo() {
  const signedIn = useAuth(select.signedIn)
```

**Comment by dai-shi** [comment]
It's almost the same as https://github.com/react-spring/zustand/issues/132#issuecomment-669619250
Having selectors outside render is better.
Yeah, this is probably the best approach with Zustand (+React).

**Comment by dai-shi** [comment]
> The new project is something different from Zustand, but with the simplicity philosophy from Zustand.

So, the new project is [jotai](https://github.com/react-spring/jotai/).

**Comment by AjaxSolutions** [comment]
> For b), we are planning a new project coming soon, and it's outside the scope of zustand.

Does this new project exist?

**Comment by woehrl01** [comment]
It's definitely possible to have computed fields, it just needs to be constructed differently to don't fall into the `Object.assign` issue. But this requirement makes the code even more readable:

```js
const [useAuth] = create((set, get) => ({
    user: {username: undefined, authLevel: 0},
   	computed: { //yes, just use a nested object, which can be easily used in `Object.assign`
       get isSignedIn: () => !!get().user.usernamecomponents?
    }
})}

//usage
const isSignedIn = useState(s => s.computed.isSignedIn)
```

**Comment by platon-rov** [comment]
@woehrl01 it won't be cached, isn't it?

**Comment by woehrl01** [comment]
@platon-rov no, it won't be cached. For caching you likely need to use a subscription to explicitly track and update dependencies.

**Comment by harry-sm** [comment]
The correct syntax for a getter according to typescript is 
```  get isSignedIn () { !!get().user.usernamecomponents? }```
not 
```  get isSignedIn: () => !!get().user.usernamecomponents?```

How does it work? @woehrl01

**Comment by evanpurkhiser** [comment]
@harry-sm you need a return then as well

```tsx
get isSignedIn() { return !!get().user.usernamecomponents? }
```

**Comment by laclance** [comment]
Can make a custom hook to update on changes
```
export default function useForceUpdate(updateOn: (state: BoundStore) => any) {
  const [update, toggleUpdate] = useState(false);

  useEffect(() => {
    const unsub = useBoundStore.subscribe(
      state => updateOn(state),
      () => toggleUpdate(!update)
    );

    return () => {
      unsub();
    };
  }, [update, updateOn]);
}
```
but its just the same work to compute the value on the fly `const isSignedIn = useBoundStore(state => !!state.user.usernamecomponents)`

**Comment by itayperry** [comment]
HI there @dai-shi,
I looked at the code you wrote here and it really helped me! Thank you :)

```javascript
const useAuth = create((set, get) => ({
    user: {username: undefined, authLevel: 0},
    signedIn: () => !!get().user.username
})}

// in components
function Foo() {
  const signedIn = useAuth(state => state.signedIn()); // ✨✨✨ This one :)

// vanilla non-reactive
const signedIn = useAuth.getState().singedIn()
// vanilla reactive
useAuth.subscribe(signedIn => console.log(signedIn), state => state.signedIn())
```
I really think this should be in the docs under a section called "computed", "derived-state", "getters" or something.. it's really helpful.
I know about the `derived-zustand` package (although it's outside the store and not in it), which is more flexible and complex, maybe it's worth mentioning too in the docs. It's just an idea of course and I don't wanna waste your time :)

**Comment by dai-shi** [comment]
>   const signedIn = useAuth(state => state.signedIn()); // ✨✨✨ This one :)

I would recommend this as a general practice.
```js
  const signedIn = useAuth(state => !!state.user.username);
```

**Comment by ossomaster** [comment]
If you are using `persist middleware` to store the data in localStorage it is necessary to remove the computed values in the configuration with partialize, example:

`{
	name: 'products',
	partialize: (state) => ({
		products: state.products,
                 // don't persist computed values
	}),
}`

**Comment by rodrigocfd** [comment]
> I'd like to share that it's definitely possible to have computed fields, it just needs to be constructed differently to don't fall into the `Object.assign` issue. But this requirement makes the code even more readable and explicit:
> 
> ```js
> const [useAuth] = create((set, get) => ({
>     user: {username: undefined, authLevel: 0},
>     computed: { //yes, just use a nested object, which can be easily used in `Object.assign`
>        get isSignedIn: () => !!get().user.usernamecomponents?
>     }
> })}
> 
> //usage
> const isSignedIn = useState(s => s.computed.isSignedIn)
> ```

While this approach is very clean, it's important to understand that there is a **performance penalty** when compared to the [custom hook approach](https://github.com/pmndrs/zustand/issues/132#issuecomment-1688161013).

Just like the custom hook, the getter function will run every time the value is requested – that is, every time the displaying component re-renders. That's fine, and that's what we expect.

However, **in addition**, this getter will also run **every time the state changes**. So, if your computation is too heavy, it's probably not a good approach. In such heavy-computation cases, you will want to memoize the computation result.

**Comment by levino** [comment]
How about using memoized selectors created with `createSelector` from [reselect](https://github.com/reduxjs/reselect)? The memoization can be shared between components somehow, I suppose. One can also [chain selectors and their memoization](https://reselect.js.org/api/createSelector).

**Comment by dai-shi** [comment]
Using memoized selectors should be good. I also develop [proxy-memoize](https://github.com/dai-shi/proxy-memoize).

**Comment by amoghkulkarnifr** [comment]
> I'd like to share that it's definitely possible to have computed fields, it just needs to be constructed differently to don't fall into the `Object.assign` issue. But this requirement makes the code even more readable and explicit:
> 
> ```js
> const [useAuth] = create((set, get) => ({
>     user: {username: undefined, authLevel: 0},
>     computed: { //yes, just use a nested object, which can be easily used in `Object.assign`
>        get isSignedIn: () => !!get().user.usernamecomponents?
>     }
> })}
> 
> //usage
> const isSignedIn = useState(s => s.computed.isSignedIn)
> ```

I'm using Zustand in TypeScript. Any idea what type `computed` state variable should be of? Can we avoid `any`?

**Comment by levino** [comment]
I really think this getter-hacking is an awful anti-pattern. This is exactly what selectors are for. For performance, use memoized selectors which can be made easily with reselect. TypeScript will also be happy.

**Comment by MarkShawn2020** [comment]
> I'd like to share that it's definitely possible to have computed fields, it just needs to be constructed differently to don't fall into the `Object.assign` issue. But this requirement makes the code even more readable and explicit:
> 
> ```js
> const [useAuth] = create((set, get) => ({
>     user: {username: undefined, authLevel: 0},
>     computed: { //yes, just use a nested object, which can be easily used in `Object.assign`
>        get isSignedIn: () => !!get().user.usernamecomponents?
>     }
> })}
> 
> //usage
> const isSignedIn = useState(s => s.computed.isSignedIn)
> ```

it would cause all the code refactored to be one level down....

**Comment by doruksahin** [comment]
Yes, getter hacking can be a problem in where calculations are expensive.
Here is another approach:

```tsx
// lets say X depends on A and B.
//before
set(draft => {
	draft.a = newA
})

//after
// first, create a new setFunction for a, since setting a won't be directly from now on
setA(newA){
	set(draft => {
		draft.a = newA
		draft.x = calculateNewX(draft.a, get().b)
	})
}

//second, check where you use set() to set a without a setter
set(draft => {
	draft.a = newA
})
to
setA(newA)
```

**Comment by doruksahin** [comment]
> I really think this getter-hacking is an awful anti-pattern. This is exactly what selectors are for. For performance, use memoized selectors which can be made easily with reselect. TypeScript will also be happy.

Yes, getter hacking can be a problem in where calculations are expensive.
Here is another approach:

`lets say X depends on A and B.`

With creating store setter function, you can calculate derived state on dependency changes

cons:
- you need to track setting `A` or `B` without setter function. (since that setter function calculates derived state)
- maybe you are not using X yet, maybe it's not mounted yet. You are still pre-calculating it (Maybe a good thing or a bad thing, depends on your business decision) 

also:
- you can check if `setA` or `setB` gets same value, if it is, return without recalculating. (I didn't add that for sake of simplicity)

```tsx
// lets say X depends on A and B.
//before
set(draft => {
	draft.a = newA
})

//after
// first, create a new setFunction for a, since setting a won't be directly from now on
setA(newA){
	set(draft => {
		draft.a = newA
		draft.x = calculateNewX(draft.a, get().b)
	})
}

//second, check where you use set() to set a without a setter
set(draft => {
	draft.a = newA
})
// to
setA(newA)
```

**Comment by yasintz** [comment]
I created [zustand-computed-state](https://www.npmjs.com/package/zustand-computed-state) library to manage computed states in Zustand using various methods. My preferred approach is the getters pattern.


```ts
const useStore = create()(
  computed((set, get) =>
    compute({
      count: 1,
      inc: () => set(prev => ({ count: prev.count + 1 })),
      dec: () => set(state => ({ count: state.count - 1 })),
      get countSq() {
        return this.count ** 2;
      },
    })
  )
);
```

**Comment by childrentime** [comment]
https://github.com/immerjs/immer/pull/941  https://github.com/immerjs/immer/issues/1015 
As mentioned in Immer's PR, using getters with immutable data is not idiomatic. I believe we should always strive to achieve getter-like effects in a 'React way'.

```ts
import { create } from "zustand";

const useAppStore = create<{
  count: number;
}>(() => ({
  count: 1,
}));

const useDoubleCount = () => {
  return useAppStore((state) => state.count * 2);
}
```

**Comment by itayperry** [comment]
Hi everyone,
Assuming my `zustand` store contains an object of type
```typescript
{ key1: {someComplicatedValue}, key2: {someComplicatedValue} ...} 
```
**or even a map** that every time I update a specific item I create a new map with `new Map(oldMap)`

Is it okay to do something like this?

```typescript

export interface StoreState {
  .
  myMap: Map<string, SomeType[]>;
  .
}
export interface StoreAction {
  .
  .
  getById: (id: string) => SomeType[] | undefined;
  .
  .
}

export const useMyStore = create<StoreState & StoreAction>()(
  (set,get) => ({
    .
    .
    .
    getById: (id) =>  get().myMap.get(id),
    .
    .
    .
  })
);
```

Can that be performant?

**Comment by mindplay-dk** [comment]
> As mentioned in Immer's PR, using getters with immutable data is not idiomatic. I believe we should always strive to achieve getter-like effects in a 'React way'.

This is the approach I ended up with as well.

The itch everyone seems to want to scratch, is your computed state ends up disjointed from your mutable state.

[`zustand-computed-state`](https://github.com/yasintz/zustand-computed-state/blob/master/README.md) looks alright, but Zustand itself really ought to have a story for this - memoized, computed state is not some sort of edge case or rare problem that most people don't have, it's just state management 101.

*Everyone* needs this - so it really needs to be supported in an idiomatic and ergonomic way.

Piling through endless issue threads and blog posts for a mediocre solution is very off putting. 😌

**Comment by mindplay-dk** [comment]
Notes to self... 🤔

> The itch everyone seems to want to scratch, is your computed state ends up disjointed from your mutable state.

The answer is "put your useStore and other related hooks in a module" - this appears to cause the least friction overall.

> Zustand itself really ought to have a story for [computed state] - memoized, computed state is not some sort of edge case or rare problem that most people don't have, it's just state management 101.

After some research, I concluded the reason Zustand doesn't have this is because... React already has that.

> Piling through endless issue threads and blog posts for a mediocre solution is very off putting. 😌

And so I decided to write [this](https://github.com/pmndrs/zustand/pull/3188) - there is really nothing surprising or off putting about just using the primitives React provides, and it seems to be how Zustand was intended to be used in the first place.

Coming from many other state management libraries, however, you're probably expecting this library to have a *story* for computed state, so it's really just a documentation issue... so I wrote this: https://github.com/pmndrs/zustand/pull/3188

There is still the issue of computed states being inaccessible outside of a component context, but tbh that's a React issue - it's probably something most React users are used to and just accept. Needing React state management outside of React components is probably going to be rare for most users.

In fact, I'd be surprised if most of the contributors in this thread (and [that one](https://github.com/pmndrs/zustand/issues/108)) aren't actually here searching for solutions mainly because there is no default, recommended approach to get you started?

That's how I ended up here for sure. 😌

**Comment by levino** [comment]
Most of the time, memoization is not needed but an overhead. Always measure before you optimise.

That being said zustand already [recommends](https://zustand.docs.pmnd.rs/getting-started/comparison#render-optimization-(vs-jotai)) using "selectors" for render optimization. Afaik the word "selector" was [coined](https://redux.js.org/usage/deriving-data-selectors) in the redux world. This lead to the creation of [`reselect`](https://github.com/reduxjs/reselect) which is an excellent library to create memoized selectors. It became so popular, it is also included in the `@reduxjs/toolkit` but can be used standalone as well. With `createSelector` and some plain old "mental gymnastics" ("How many instances of this will be created?", "Will there actually be a cache hit?", there is no way around painstakingly posing and answering these questions. ChatGPT and friends will not help you here! This is where the "magic" happens. Often the selector instance should be a static component ref so it caches the rerenders of this component.) you can optimize it all. By just knowing that when a "selector" returns the same value (as in `Object.is` says it is the same), you skip a rerender. That is imo all there is to know and there is also no need to "explain" more or make "examples". All the optimization examples I have seen make very little sense because they construct some "expensive calculation" out of nowhere, which rarely happens. In my experience it is not that a calculation is expensive but that a state change triggers so many rerenders of (sub-)components that the app becomes sluggish. So the important part is not the "memoization" but the "memoization" is a means to an end: Bail and do not rerender all of the components (think of 50 like buttons and you only want to increase the likes of one of these by updating the array of all currently visible "posts").

Also this bailing is imo the "secret sauce" of zustand which makes it even better than redux for me.

**Comment by dai-shi** [comment]
Not sure if this is the right place to continue the discussion, but:
- Using the hook is a recommended solution for React.
- For vanilla, I made an experimental library `derive-zustand` but it didn't get much traction.

**Comment by levino** [comment]
Which hook?

**Comment by dai-shi** [comment]
```js
import { create } from 'zustand';
const useFooStore = create(...);
//        ^ this hook
```
Please open a new discussion for further discussion.
